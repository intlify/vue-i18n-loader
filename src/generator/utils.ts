import { isObject, isString } from '@intlify/shared'
import { VisitorKeys } from 'eslint-visitor-keys'
import { parseForESLint, VisitorKeys as VISITOR_KEYS } from 'yaml-eslint-parser'

import type { YAMLProgram, YAMLNode } from 'yaml-eslint-parser/lib/ast'

export interface Visitor<N> {
  visitorKeys?: VisitorKeys
  enterNode(node: N, parent: N | null): void
  leaveNode(node: N, parent: N | null): void
}

export function parseYAML(code: string, options = {}): YAMLProgram {
  const parserOptions = Object.assign(
    {
      filePath: '<input>',
      ecmaVersion: 2019
    },
    options,
    {
      loc: true,
      range: true,
      raw: true,
      tokens: true,
      comment: true,
      eslintVisitorKeys: true,
      eslintScopeManager: true
    }
  )

  return parseForESLint(code, parserOptions).ast
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function fallbackKeysFilter(this: any, key: string) {
  let value = null
  return (
    key !== 'comments' &&
    key !== 'sourceType' &&
    key !== 'loc' &&
    key !== 'parent' &&
    key !== 'range' &&
    key !== 'tokens' &&
    (value = this[key]) !== null &&
    isObject(value) &&
    (isString(value.type) || Array.isArray(value))
  )
}

function getFallbackKeys(node: YAMLNode): string[] {
  return Object.keys(node).filter(fallbackKeysFilter, node)
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function isNode(x: any): x is YAMLNode {
  return x !== null && isObject(x) && isString(x.type)
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/explicit-module-boundary-types
export function* getNodes(node: any, key: string): IterableIterator<YAMLNode> {
  const child = node[key]
  if (Array.isArray(child)) {
    for (const c of child) {
      if (isNode(c)) {
        yield c
      }
    }
  } else if (isNode(child)) {
    yield child
  }
}

function getKeys(node: YAMLNode, visitorKeys?: VisitorKeys) {
  const keys = (visitorKeys || VISITOR_KEYS)[node.type] || getFallbackKeys(node)
  return keys.filter((key: string) => !getNodes(node, key).next().done)
}

function traverse(
  node: YAMLNode,
  parent: YAMLNode | null,
  visitor: Visitor<YAMLNode>
) {
  visitor.enterNode(node, parent)
  const keys = getKeys(node, visitor.visitorKeys)
  for (const key of keys) {
    for (const child of getNodes(node, key)) {
      traverse(child, node, visitor)
    }
  }
  visitor.leaveNode(node, parent)
}

export function traverseNodes(
  node: YAMLNode,
  visitor: Visitor<YAMLNode>
): void {
  traverse(node, null, visitor)
}
