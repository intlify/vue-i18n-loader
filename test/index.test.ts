import { bundleAndRun } from './utils'
import { MessageFunction, baseCompile } from 'vue-i18n'
import prettier from 'prettier'

test('basic', async () => {
  const { module } = await bundleAndRun('basic.vue')
  expect(module.__i18n).toMatchSnapshot()
})

test('special characters', async () => {
  const { module } = await bundleAndRun('special-char.vue')
  expect(module.__i18n).toMatchSnapshot()
})

test('multiple', async () => {
  const { module } = await bundleAndRun('multiple.vue')
  expect(module.__i18n).toMatchSnapshot()
})

test('import', async () => {
  const { module } = await bundleAndRun('import.vue')
  expect(module.__i18n).toMatchSnapshot()
})

test('locale attr', async () => {
  const { module } = await bundleAndRun('locale.vue')
  expect(module.__i18n).toMatchSnapshot()
})

test('locale attr and basic', async () => {
  const { module } = await bundleAndRun('locale-mix.vue')
  expect(module.__i18n).toMatchSnapshot()
})

test('locale attr and import', async () => {
  const { module } = await bundleAndRun('locale-import.vue')
  expect(module.__i18n).toMatchSnapshot()
})

test('yaml', async () => {
  const { module } = await bundleAndRun('yaml.vue')
  expect(module.__i18n).toMatchSnapshot()
})

test('json5', async () => {
  const { module } = await bundleAndRun('json5.vue')
  expect(module.__i18n).toMatchSnapshot()
})

test('pre compile', async () => {
  const options: prettier.Options = { parser: 'babel' }
  const { module } = await bundleAndRun('compile.vue', {
    preCompile: true
  })
  const { functions } = module.__i18n()
  for (const [key, value] of Object.entries(functions)) {
    const msg = value as MessageFunction
    const data = JSON.parse(key)
    const result = baseCompile(data.s, { mode: 'arrow' })
    expect(prettier.format(msg.toString(), options)).toMatch(
      prettier.format(result.code, options)
    )
  }
})
