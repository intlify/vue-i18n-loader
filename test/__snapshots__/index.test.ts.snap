// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic 1`] = `
Array [
  "{\\"en\\":{\\"hello\\":\\"hello world!\\"}}",
]
`;

exports[`import 1`] = `
Array [
  "{\\"en\\":{\\"hello\\":\\"hello world!\\"}}",
]
`;

exports[`json5 1`] = `
Array [
  "{\\"en\\":{\\"hello\\":\\"hello world!\\"}}",
]
`;

exports[`locale attr 1`] = `
Array [
  "{\\"ja\\":{\\"hello\\":\\"こんにちは、世界！\\"}}",
]
`;

exports[`locale attr and basic 1`] = `
Array [
  "{\\"en\\":{\\"hello\\":\\"hello world!\\"}}",
  "{\\"ja\\":{\\"hello\\":\\"こんにちは、世界！\\"}}",
]
`;

exports[`locale attr and import 1`] = `
Array [
  "{\\"en\\":{\\"hello\\":\\"hello world!\\"}}",
]
`;

exports[`multiple 1`] = `
Array [
  "{\\"en\\":{\\"hello\\":\\"hello world!\\"}}",
  "{\\"ja\\":{\\"hello\\":\\"こんにちは、世界！\\"}}",
]
`;

exports[`special characters 1`] = `
Array [
  "{\\"en\\":{\\"hello\\":\\"hello
great	\\"world\\"\\"}}",
]
`;

exports[`yaml 1`] = `
Array [
  "{\\"en\\":{\\"hello\\":\\"hello world!\\"}}",
  "{\\"ja\\":{\\"hello\\":\\"こんにちは、世界！\\"}}",
]
`;
